infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -272,18 +290,29 @@ public void read(CorfuPayloadMsg<ReadRequest> msg, ChannelHandlerContext ctx, IS
         boolean cacheable = msg.getPayload().isCacheReadResult();
         log.trace("read: {}, cacheable: {}", msg.getPayload().getAddress(), cacheable);
 
-        ReadResponse rr = new ReadResponse();
+        Lock lock = resetLock.readLock();
+        lock.lock();
+
         try {
-            ILogData logData = dataCache.get(address, cacheable);
-            if (logData == null) {
-                rr.put(address, LogData.getEmpty(address));
-            } else {
-                rr.put(address, (LogData) logData);
+            if (msg.getEpoch() != serverContext.getServerEpoch()) {
+                throw new WrongEpochException(serverContext.getServerEpoch());
```

Comment:
Shouldn't this be based on serverContext.getLogUnitEpochWaterMark()?

Reply:
What's the motivation for that?

Reply:
The motivation is "consistency". The batch processor will fail requests based on `LogUnitEpochWaterMark` and not `ServerEpoch`

Reply:
fixed